executor/insert_common.go

```diff
@@ -528,9 +588,141 @@ func insertRowsFromSelect(ctx context.Context, base insertCommon) error {
 		memTracker.Consume(-memUsageOfExtraCols)
 		memTracker.Consume(-chkMemUsage)
 	}
+	if etlMode {
+		close(dataCh)
+	}
 	return nil
 }
 
+// CreateSession will be assigned in session package
+var CreateSession func(ctx sessionctx.Context) (sessionctx.Context, error)
+
+// CloseSession will be assigned in session package
+var CloseSession func(ctx sessionctx.Context)
+
+func createFromSelectWorkers(ctx context.Context, base insertCommon, concurrency int) []insertCommon {
+	insert, ok := base.(*InsertExec)
+	if !ok {
+		return nil
+	}
+	inserts := make([]insertCommon, 0, concurrency)
+	for i := 0; i < concurrency; i++ {
+		se, err := CreateSession(insert.ctx)
+		if err != nil {
+			logutil.Logger(ctx).Error("create session failed", zap.Error(err))
+			return nil
+		}
+		if len(insert.fkCascades) > 0 || len(insert.fkChecks) > 0 {
+			logutil.Logger(ctx).Error("fk not supported")
+			return nil
+		}
+		baseExec := newBaseExecutor(se, insert.schema, insert.id, insert.children...)
+		baseExec.runtimeStats = &execdetails.BasicRuntimeStats{}
+		insertValues := &InsertValues{
+			baseExecutor:              baseExec,
+			SelectExec:                insert.SelectExec,
+			Table:                     insert.Table,
+			Columns:                   insert.Columns,
+			Lists:                     insert.Lists,
+			SetList:                   insert.SetList,
+			GenExprs:                  insert.GenExprs,
+			insertColumns:             insert.insertColumns,
+			colDefaultVals:            insert.colDefaultVals,
+			evalBuffer:                chunk.MutRowFromTypes(insert.evalBufferTypes),
+			evalBufferTypes:           insert.evalBufferTypes,
+			allAssignmentsAreConstant: insert.allAssignmentsAreConstant,
+			hasRefCols:                insert.hasRefCols,
+			hasExtraHandle:            insert.hasExtraHandle,
+			lazyFillAutoID:            insert.lazyFillAutoID,
+			memTracker:                insert.memTracker,
+			rowLen:                    insert.rowLen,
+			stats:                     nil,
+			isLoadData:                insert.isLoadData,
+			fkChecks:                  nil,
+			fkCascades:                nil,
+		}
+		insertValues.collectRuntimeStatsEnabled()
+		newInsert := &InsertExec{
+			InsertValues:   insertValues,
+			OnDuplicate:    insert.OnDuplicate,
+			evalBuffer4Dup: insert.evalBuffer4Dup,
+			curInsertVals:  insert.curInsertVals,
+			row4Update:     insert.row4Update,
+			Priority:       insert.Priority,
+		}
+		inserts = append(inserts, newInsert)
+	}
+	return inserts
+}
+
+func insertRowsFromSelectWorker(ctx context.Context, base insertCommon, batchSize int, dataCh <-chan [][]types.Datum,
+	wg *sync.WaitGroup, globalAffectedRows *uint64) {
+	e := base.insertCommon()
+	se := e.ctx
+	sessVars := se.GetSessionVars()
+	// assert no conflict, but auto retry
+	sessVars.TxnMode = "OPTIMISTIC"
+	sessVars.DisableTxnAutoRetry = false
+
+	rows := make([][]types.Datum, 0, batchSize)
+	extraColsInSel := make([][]types.Datum, 0, batchSize)
```

Comment:
This seems not to be used, we could remove it.

Reply:
Used only by on duplicate update.

Reply:
Skip it when there is no on duplicate clause.